Lift distribution post-processor

sduess opened the pull request:
Lift distribution post-processor was failing due to missing adjustments to previous SHARPy changes. The post-processor now reads aerodynamic forces from aero timestep and adjusts the lift force at nodes shared from two different surfaces (e.g. at symmetry plane). Adjusted lift force along with x-,y-, and z-coordinates of leading are exported to txt format for lift distribution evaluations. Code has no impact on any other SHARPy code.

ArturoMS13 commented on sharpy/postproc/liftdistribution.py:
```diff
@@ -1,19 +1,17 @@
 import sharpy.utils.cout_utils as cout
-from sharpy.utils.settings import str2bool
 from sharpy.utils.solver_interface import solver, BaseSolver
 import sharpy.utils.settings as settings
```
```suggestion
import sharpy.aero.utils.mapping as mapping
```

ArturoMS13 replied:
```suggestion
import sharpy.utils.settings as settings
import sharpy.aero.utils.mapping as mapping
```

ArturoMS13 commented on sharpy/postproc/liftdistribution.py:
```diff
@@ -49,35 +51,41 @@ def run(self, online=False):
-
-        norm = 1.0
-
+        tstep = self.data.aero.timestep_info[self.data.ts]
```
```suggestion
       aero_tstep = self.data.aero.timestep_info[self.data.ts]
       struct_tstep = self.data.structure.timestep_info[self.data.ts]
```

ArturoMS13 commented on sharpy/postproc/liftdistribution.py:
```diff
@@ -49,35 +51,41 @@ def run(self, online=False):
+        node_counter = 0
+        for i_surf in range(tstep.n_surf):
+            n_dim, n_m, n_n = tstep.zeta[i_surf].shape
+            forces = tstep.forces[i_surf] + tstep.dynamic_forces[i_surf]
```
```suggestion
        forces = mapping.aero2struct_force_mapping(
            aero_tstep.forces,
            self.data.aero.struct2aero_mapping,
            aero_tstep.zeta,
            structural_tstep.pos,
            structural_tstep.psi,
            self.data.structure.node_master_elem,
            self.data.structure.connectivities,
            structural_tstep.cag(),
            self.data.aero.aero_dict)
        forces += mapping.aero2struct_force_mapping(
            aero_tstep.dynamic_forces,
            self.data.aero.struct2aero_mapping,
            aero_tstep.zeta,
            structural_tstep.pos,
            structural_tstep.psi,
            self.data.structure.node_master_elem,
            self.data.structure.connectivities,
            structural_tstep.cag(),
            self.data.aero.aero_dict)

```

ArturoMS13 commented on sharpy/postproc/liftdistribution.py:
```diff
@@ -49,35 +51,41 @@ def run(self, online=False):
-        norm = 1.0
-
+        tstep = self.data.aero.timestep_info[self.data.ts]
+        k_total = 0
```
```suggestion

```

ArturoMS13 commented on sharpy/postproc/liftdistribution.py:
```diff
@@ -49,35 +51,41 @@ def run(self, online=False):
+            k_total += n_n
+            
+        # get aero forces
+        lift_distribution = np.zeros((k_total, 5))
```
```suggestion
        lift_distribution = np.zeros((self.data.structure.num_node, 5))
```

ArturoMS13 commented on sharpy/postproc/liftdistribution.py:
```diff
@@ -49,35 +51,41 @@ def run(self, online=False):
+            
+        # get aero forces
+        lift_distribution = np.zeros((k_total, 5))
+        node_counter = 0
+        for i_surf in range(tstep.n_surf):
+            n_dim, n_m, n_n = tstep.zeta[i_surf].shape
```
```suggestion

```

ArturoMS13 commented on sharpy/postproc/liftdistribution.py:
```diff
@@ -49,35 +51,41 @@ def run(self, online=False):
+        for i_surf in range(tstep.n_surf):
+            n_dim, n_m, n_n = tstep.zeta[i_surf].shape
+            forces = tstep.forces[i_surf] + tstep.dynamic_forces[i_surf]
             abs_forces = np.zeros((n_m, n_n))
```
```suggestion
            abs_forces = np.zeros((self.data.structure.num_node))
```

ArturoMS13 commented:
Hi Stef, 
thank you for the very interesting feature. 
I have commit some changes to keep the function compatible with any kind of structure that can be created in sharpy.
Can you have a look at it? 
Let me know if I forgot something

ngoiz commented:
Hi all, and nicely done on this first PR, Stef!

One thing I noticed is that the settings for this solver use what was the old convention (`self.settings_default` etc. as an instance attribute defined in the class' `__init__()`). To make the story short, back when I did the documentation builder settings defaults/types... need to be in given as class attributes such that they are included in the docs (the tables that appear online are self generated and require this). So if you can move them to class attributes and include the docstring modification line that'd be great (you can check any solver that has docs online to see how it's done).

Then, another thought that comes to mind is whether we'd like to merge this postproc with `AeroForcesCalculator`. It seems like both are doing the similar things and a lot of code is very similar. I was just updating it the other day to compute moments as well about the A frame (it is in `dev_moments`) so maybe we can bring all together to make a more comprehensive solver? Thoughts?

sduess commented:
Hi guys! 
I have updated the lift distribution post-processor. I have adapted some of your suggestions and improved the code. The new setting convention is used now and you can export the lift coefficient for which a panel area calculation was needed and implemented. 
For now, I would not merge the postprocessor with the AeroforcesCalculator, as they are highly customized to their respective tasks and not easy to merge.

Let me know what you think!

ngoiz commented on sharpy/postproc/liftdistribution.py:
```diff
@@ -1,79 +1,131 @@
+        for inode in range(N_nodes):
+            if self.data.aero.aero_dict['aero_node'][inode]:
+                # transform forces from B to A frame
+                lift_distribution[inode,3]=np.dot(rot.T, forces[inode, :3])[2]  # lift force
```
Hmm I believe this will not always be the lift force. In the case, say the fluid is rotated but the wing is straight (i.e. A and G coincident) you will not be capturing the perpendicular to the free stream.

How @ArturoMS13 implemented this was to find the free stream direction at the section and project the force onto it (this would be the drag), then, the remainder of the force is assumed to be the lift (there would be a side force component particularly near tips but it makes the calculation simpler). 

If you have a look at the Polars file you may it is done there if you'd like to implement it

ngoiz replied:
Also, if I'm not mistaken `forces` is in B frame. To get them to `A` you'll need the local CRV to find `cab` and then you can use `rot.T` to get them onto G.

This is always so confusing......it's been years for me and still have doubts. I put a little function in `algebra` called `get_transformation()` that gets you the projection matrix based on the string you give it.

ngoiz commented on sharpy/utils/algebra.py:
```diff
@@ -269,7 +269,19 @@ def angle_between_vector_and_plane(vector, plane_normal):
 
-
+def panel_area(A, B, C, D):
+    # calculates the area of a quadrilateral panel from the corner 
+    # points A,B,C, and D using Bertschneider's formula                 
```
```suggestion
    """
    Calculates the area of a quadrilateral panel from the corner 
    points A,B,C, and D using Bertschneider's formula
    
    Args:
        A (np.ndarray): Coordinates of point 1
        B (np.ndarray): Coordinates of point 2
        C (np.ndarray): Coordinates of point 3
        D (np.ndarray): Coordinates of point 4

    Returns:
        float: Area of quadrilateral panel
    """      
```

By adding the little description as a docstring it will automatically appear in the algebra module documentation online.

The formatting for the docstring follows the Google convention

ngoiz reviewed (changes requested):
Looking good Stef! I left a few comments on getting the lift but I do agree with you in that it is best keeping this one separate from `AeroForcesCalculator`.

Let me know if you have any questions about the comments

ArturoMS13 commented on sharpy/postproc/liftdistribution.py:
```diff
@@ -80,44 +80,44 @@ def lift_distribution(self, struct_tstep, aero_tstep):
-                                                  / (0.5 * self.settings['rho'] * np.linalg.norm(
-                        urel) ** 2 * span * chord)  # strip_area[i_surf][local_node])
+                    # Get non-dimensional spanwise coordinate y/s
+                    lift_distribution[inode, 4] = lift_distribution[inode, 1]/max(abs(aero_tstep.zeta[i_surf][1,0,:]))
```
```suggestion
                    lift_distribution[inode, 4] = lift_distribution[inode, 1]/span
```

ArturoMS13 reviewed (approved):
Thank you for the contribution!

ngoiz reviewed (approved):
Many thanks!
